File: Python/0002.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        n1, n2 = l1, l2
        
        while n1 and n2:
            n1.val += n2.val
            if not n1.next:
                n1.next = n2.next
                break
            n1, n2 = n1.next, n2.next

        carry = 0
        n1 = l1
        while n1:
            n1.val += carry
            carry = n1.val // 10
            n1.val %= 10
            if not n1.next and carry:
                n1.next = ListNode(carry)
                break
            n1 = n1.next

        return l1

File: C++/0002.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0, firstNode = 1;
        auto res = l1, curr = l1;
        while(l1 || l2 || carry != 0)
        {
            if(l1)
                carry += l1->val, l1 = l1->next;
            if(l2)
                carry += l2->val, l2 = l2->next;
            auto newNode = new ListNode(carry % 10);
            if(firstNode)
                res = curr = newNode, --firstNode;
            else
                curr->next = newNode, curr = curr->next;
            carry /= 10;
        }
        return res;
    }
};